File: employee_scheduler.js
class EmployeeScheduler {
    constructor() {
        // Using an object here for employees with their shift preferences
        this.employees = {};

        // Using an object here for schedule with days and shifts
        this.schedule = {
            "Monday": { "morning": [], "afternoon": [], "evening": [] },
            "Tuesday": { "morning": [], "afternoon": [], "evening": [] },
            "Wednesday": { "morning": [], "afternoon": [], "evening": [] },
            "Thursday": { "morning": [], "afternoon": [], "evening": [] },
            "Friday": { "morning": [], "afternoon": [], "evening": [] },
            "Saturday": { "morning": [], "afternoon": [], "evening": [] },
            "Sunday": { "morning": [], "afternoon": [], "evening": [] }
        };

        // Using an object here to track the number of days each employee has worked
        this.workdays = {};

        // Using an object here to track the number of shifts assigned to each employee
        this.shiftsAssigned = {};
    }

    // Input and storage of employee names and their shift preferences and priority rankings into the objects
    addEmployeePreferences(name, preferences) {
        this.employees[name] = preferences;
        this.workdays[name] = 0;
        this.shiftsAssigned[name] = 0;
    }

    // Assign shifts for employees based on preferences and scheduling rules
    assignShifts() {
        for (let day in this.schedule) {
            // Create a set of unassigned employees
            let unassigned = new Set(Object.keys(this.employees));

            for (let shift of ["morning", "afternoon", "evening"]) {
                // Find employees who prefer this shift and have worked less than 5 days
                let preferredEmployees = Array.from(unassigned).filter(e => this.employees[e][day].includes(shift) && this.workdays[e] < 5);

                if (preferredEmployees.length >= 2) {
                    // Sort (to give everyone an equal chance) employees by the number of shifts assigned and their preference ranking for this shift
                    preferredEmployees.sort((a, b) => this.shiftsAssigned[a] - this.shiftsAssigned[b] || this.employees[a][day].indexOf(shift) - this.employees[b][day].indexOf(shift));

                    // Assign the first 2 preferred employees to the shift
                    this.assignToShift(day, shift, preferredEmployees.slice(0, 2), unassigned);
                } else {
                    // Find other available employees if preferred employees are less than 2
                    let availableEmployees = Array.from(unassigned).filter(e => this.workdays[e] < 5 && !preferredEmployees.includes(e));

                    // Sort available employees by the number of shifts assigned
                    availableEmployees.sort((a, b) => this.shiftsAssigned[a] - this.shiftsAssigned[b]);

                    // Assign the preferred employees and the first available employees to fill the shift
                    this.assignToShift(day, shift, preferredEmployees.concat(availableEmployees.slice(0, 2 - preferredEmployees.length)), unassigned);
                }

                // Ensure at least 2 employees per shift
                if (this.schedule[day][shift].length < 2) {
                    // Find additional employees who have worked less than 5 days and are not already assigned to this shift
                    let additionalEmployees = Object.keys(this.employees).filter(e => this.workdays[e] < 5 && !this.schedule[day]["morning"].includes(e) && !this.schedule[day]["afternoon"].includes(e) && !this.schedule[day]["evening"].includes(e));

                    // Sort additional employees by the number of shifts assigned
                    additionalEmployees.sort((a, b) => this.shiftsAssigned[a] - this.shiftsAssigned[b]);

                    // Assign the first additional employees to fill the shift
                    this.assignToShift(day, shift, additionalEmployees.slice(0, 2 - this.schedule[day][shift].length), unassigned);
                }
            }
        }
    }

    // Assign the shift to the employees in assigned and update their workdays and shifts assigned
    assignToShift(day, shift, employees, unassigned) {
        for (let e of employees) {
            if (unassigned.has(e)) {
                this.schedule[day][shift].push(e);
                this.workdays[e]++;
                this.shiftsAssigned[e]++;
                unassigned.delete(e);
            }
        }
    }

    // Logic to detect and resolve conflicts
    resolveConflicts() {
        for (let day in this.schedule) {
            for (let shift in this.schedule[day]) {
                if (this.schedule[day][shift].length > 2) {
                    // If more than 2 employees are assigned to a shift, reassign the excess employees
                    let excessEmployees = this.schedule[day][shift].slice(2);

                    this.schedule[day][shift] = this.schedule[day][shift].slice(0, 2);

                    for (let e of excessEmployees) {
                        for (let altShift of ["morning", "afternoon", "evening"]) {
                            if (this.schedule[day][altShift].length < 2) {
                                this.schedule[day][altShift].push(e);
                                break;
                            }
                        }
                    }
                }
            }
        }
    }

    // Print the weekly schedule
    displaySchedule() {
        for (let day in this.schedule) {
            console.log(`${day}:`);
            for (let shift in this.schedule[day]) {
                console.log(`  ${shift.charAt(0).toUpperCase() + shift.slice(1)}: ${this.schedule[day][shift].join(", ") || "No employees assigned"}`);
            }
            console.log("-".repeat(30));
        }
    }
}

// Example usage
const scheduler = new EmployeeScheduler();

const employeeData = {
    "Employee A": {
        "Monday": ["morning", "afternoon", "evening"],
        "Tuesday": ["morning", "afternoon", "evening"],
        "Wednesday": ["morning", "afternoon", "evening"],
        "Thursday": ["morning", "afternoon", "evening"],
        "Friday": ["morning", "afternoon", "evening"],
        "Saturday": ["morning", "afternoon", "evening"],
        "Sunday": ["morning", "afternoon", "evening"]
    },
    "Employee B": {
        "Monday": ["afternoon", "morning", "evening"],
        "Tuesday": ["afternoon", "morning", "evening"],
        "Wednesday": ["afternoon", "morning", "evening"],
        "Thursday": ["afternoon", "morning", "evening"],
        "Friday": ["afternoon", "morning", "evening"],
        "Saturday": ["afternoon", "morning", "evening"],
        "Sunday": ["afternoon", "morning", "evening"]
    },
    "Employee C": {
        "Monday": ["evening", "morning", "afternoon"],
        "Tuesday": ["evening", "morning", "afternoon"],
        "Wednesday": ["evening", "morning", "afternoon"],
        "Thursday": ["evening", "morning", "afternoon"],
        "Friday": ["evening", "morning", "afternoon"],
        "Saturday": ["evening", "morning", "afternoon"],
        "Sunday": ["evening", "morning", "afternoon"]
    },
    "Employee D": {
        "Monday": ["morning", "evening", "afternoon"],
        "Tuesday": ["morning", "evening", "afternoon"],
        "Wednesday": ["morning", "evening", "afternoon"],
        "Thursday": ["morning", "evening", "afternoon"],
        "Friday": ["morning", "evening", "afternoon"],
        "Saturday": ["morning", "evening", "afternoon"],
        "Sunday": ["morning", "evening", "afternoon"]
    },
    "Employee E": {
        "Monday": ["afternoon", "evening", "morning"],
        "Tuesday": ["afternoon", "evening", "morning"],
        "Wednesday": ["afternoon", "evening", "morning"],
        "Thursday": ["afternoon", "evening", "morning"],
        "Friday": ["afternoon", "evening", "morning"],
        "Saturday": ["afternoon", "evening", "morning"],
        "Sunday": ["afternoon", "evening", "morning"]
    },
    "Employee F": {
        "Monday": ["morning", "afternoon", "evening"],
        "Tuesday": ["morning", "afternoon", "evening"],
        "Wednesday": ["morning", "afternoon", "evening"],
        "Thursday": ["morning", "afternoon", "evening"],
        "Friday": ["morning", "afternoon", "evening"],
        "Saturday": ["morning", "afternoon", "evening"],
        "Sunday": ["morning", "afternoon", "evening"]
    },
    "Employee G": {
        "Monday": ["afternoon", "morning", "evening"],
        "Tuesday": ["afternoon", "morning", "evening"],
        "Wednesday": ["afternoon", "morning", "evening"],
        "Thursday": ["afternoon", "morning", "evening"],
        "Friday": ["afternoon", "morning", "evening"],
        "Saturday": ["afternoon", "morning", "evening"],
        "Sunday": ["afternoon", "morning", "evening"]
    },
    "Employee H": {
        "Monday": ["evening", "morning", "afternoon"],
        "Tuesday": ["evening", "morning", "afternoon"],
        "Wednesday": ["evening", "morning", "afternoon"],
        "Thursday": ["evening", "morning", "afternoon"],
        "Friday": ["evening", "morning", "afternoon"],
        "Saturday": ["evening", "morning", "afternoon"],
        "Sunday": ["evening", "morning", "afternoon"]
    },
    "Employee I": {
        "Monday": ["morning", "evening", "afternoon"],
        "Tuesday": ["morning", "evening", "afternoon"],
        "Wednesday": ["morning", "evening", "afternoon"],
        "Thursday": ["morning", "evening", "afternoon"],
        "Friday": ["morning", "evening", "afternoon"],
        "Saturday": ["morning", "evening", "afternoon"],
        "Sunday": ["morning", "evening", "afternoon"]
    },
    "Employee J": {
        "Monday": ["afternoon", "evening", "morning"],
        "Tuesday": ["afternoon", "evening", "morning"],
        "Wednesday": ["afternoon", "evening", "morning"],
        "Thursday": ["afternoon", "evening", "morning"],
        "Friday": ["afternoon", "evening", "morning"],
        "Saturday": ["afternoon", "evening", "morning"],
        "Sunday": ["afternoon", "evening", "morning"]
    }
};

for (let name in employeeData) {
    scheduler.addEmployeePreferences(name, employeeData[name]);
}

scheduler.assignShifts();
scheduler.resolveConflicts();
scheduler.displaySchedule();

File: employee_scheduler.py
from collections import defaultdict

class EmployeeScheduler:
    def __init__(self):
        # Using a dictionary here for employees with their shift preference
        self.employees = {}

        # Using a dictionary here for schedule with days and shifts
        self.schedule = {day: {"morning": [], "afternoon": [], "evening": []} for day in
                         ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]}

        # Using a dictionary here to track the number of days each employee has worked
        self.workdays = defaultdict(int)

        # Using a dictionary here to track the number of shifts assigned to each employee
        self.shifts_assigned = defaultdict(int)

    # Input and storage of employee names and their shift preferences and priority rankings
    def add_employee_preferences(self, name, preferences):
        # Add an employee with their shift preferences and priority rankings
        self.employees[name] = preferences

    # Assign shifts for employees based on preferences and scheduling rules
    def assign_shifts(self):
        for day in self.schedule:
            # Create a set of unassigned employees (fill this with all employees as we haven't started assigning yet)
            unassigned = set(self.employees.keys())

            for shift in ["morning", "afternoon", "evening"]:
                # Find employees who prefer this shift and have worked less than 5 days (more than which they cannot be scheduled)
                preferred_employees = [e for e in unassigned if shift in self.employees[e][day] and self.workdays[e] < 5]

                if len(preferred_employees) >= 2:
                    # Sort (to give everyone an equal chance) employees by the number of shifts assigned and their preference ranking for this shift
                    preferred_employees.sort(key=lambda e: (self.shifts_assigned[e], self.employees[e][day].index(shift)))

                    # Assign the first 2 preferred employees to the shift
                    assigned = preferred_employees[:2]

                else:
                    # Find other available employees if preferred employees are less than 2
                    available_employees = [e for e in unassigned if self.workdays[e] < 5 and e not in preferred_employees]

                    # Sort available employees by the number of shifts assigned
                    available_employees.sort(key=lambda e: self.shifts_assigned[e])

                    # Assign the preferred employees and the first available employees to fill the shift
                    assigned = preferred_employees + available_employees[:2 - len(preferred_employees)]

                # Assign the shift to the employees in assigned and update their workdays and shifts assigned
                for e in assigned:
                    if e in unassigned:
                        self.schedule[day][shift].append(e)
                        self.workdays[e] += 1
                        self.shifts_assigned[e] += 1
                        unassigned.remove(e)

                # Ensure at least 2 employees per shift
                # If fewer than 2 employees are available for a shift
                # randomly assign additional employees who have not worked 5 days yet
                if len(self.schedule[day][shift]) < 2:
                    # Find additional employees who have not worked 5 days and are not already assigned to a shift on this day
                    additional_employees = [e for e in self.employees if self.workdays[e] < 5 and e not in self.schedule[day]["morning"] and e not in self.schedule[day]["afternoon"] and e not in self.schedule[day]["evening"]]

                    # Sort additional employees by the number of shifts assigned
                    additional_employees.sort(key=lambda e: self.shifts_assigned[e])

                    # Calculate the number of additional employees needed to fill the shift
                    additional_needed = 2 - len(self.schedule[day][shift])

                    # Assign the first additional employees to fill the shift
                    additional_assigned = additional_employees[:additional_needed]

                    # Assign the shift to the employees in additional_assigned and update their workdays and shifts assigned
                    for e in additional_assigned:
                        self.schedule[day][shift].append(e)
                        self.workdays[e] += 1
                        self.shifts_assigned[e] += 1

    # Logic to detect and resolve conflicts
    def resolve_conflicts(self):
        # Ensure employees get at least one shift close to their preference if it is unavailable
        for day in self.schedule:
            for shift in self.schedule[day]:
                if len(self.schedule[day][shift]) > 2:
                    # If more than 2 employees are assigned to a shift, reassign the excess employees
                    excess_employees = self.schedule[day][shift][2:]
                    self.schedule[day][shift] = self.schedule[day][shift][:2]
                    for e in excess_employees:
                        for alt_shift in ["morning", "afternoon", "evening"]:
                            if len(self.schedule[day][alt_shift]) < 2:
                                self.schedule[day][alt_shift].append(e)
                                break

    def display_schedule(self):
        # Prints the weekly schedule
        for day, shifts in self.schedule.items():
            print(f"{day}: ")
            for shift, employees in shifts.items():
                print(f"  {shift.capitalize()}: {', '.join(employees) if employees else 'No employees assigned'}")
            print("-" * 30)

# Example usage
scheduler = EmployeeScheduler()

# Add employees with their shift preferences and priority rankings
employee_data = {
    "Employee A": {
        "Monday": ["morning", "afternoon", "evening"],
        "Tuesday": ["morning", "afternoon", "evening"],
        "Wednesday": ["morning", "afternoon", "evening"],
        "Thursday": ["morning", "afternoon", "evening"],
        "Friday": ["morning", "afternoon", "evening"],
        "Saturday": ["morning", "afternoon", "evening"],
        "Sunday": ["morning", "afternoon", "evening"]
    },
    "Employee B": {
        "Monday": ["afternoon", "morning", "evening"],
        "Tuesday": ["afternoon", "morning", "evening"],
        "Wednesday": ["afternoon", "morning", "evening"],
        "Thursday": ["afternoon", "morning", "evening"],
        "Friday": ["afternoon", "morning", "evening"],
        "Saturday": ["afternoon", "morning", "evening"],
        "Sunday": ["afternoon", "morning", "evening"]
    },
    "Employee C": {
        "Monday": ["evening", "morning", "afternoon"],
        "Tuesday": ["evening", "morning", "afternoon"],
        "Wednesday": ["evening", "morning", "afternoon"],
        "Thursday": ["evening", "morning", "afternoon"],
        "Friday": ["evening", "morning", "afternoon"],
        "Saturday": ["evening", "morning", "afternoon"],
        "Sunday": ["evening", "morning", "afternoon"]
    },
    "Employee D": {
        "Monday": ["morning", "evening", "afternoon"],
        "Tuesday": ["morning", "evening", "afternoon"],
        "Wednesday": ["morning", "evening", "afternoon"],
        "Thursday": ["morning", "evening", "afternoon"],
        "Friday": ["morning", "evening", "afternoon"],
        "Saturday": ["morning", "evening", "afternoon"],
        "Sunday": ["morning", "evening", "afternoon"]
    },
    "Employee E": {
        "Monday": ["afternoon", "evening", "morning"],
        "Tuesday": ["afternoon", "evening", "morning"],
        "Wednesday": ["afternoon", "evening", "morning"],
        "Thursday": ["afternoon", "evening", "morning"],
        "Friday": ["afternoon", "evening", "morning"],
        "Saturday": ["afternoon", "evening", "morning"],
        "Sunday": ["afternoon", "evening", "morning"]
    },
    "Employee F": {
        "Monday": ["morning", "afternoon", "evening"],
        "Tuesday": ["morning", "afternoon", "evening"],
        "Wednesday": ["morning", "afternoon", "evening"],
        "Thursday": ["morning", "afternoon", "evening"],
        "Friday": ["morning", "afternoon", "evening"],
        "Saturday": ["morning", "afternoon", "evening"],
        "Sunday": ["morning", "afternoon", "evening"]
    },
    "Employee G": {
        "Monday": ["afternoon", "morning", "evening"],
        "Tuesday": ["afternoon", "morning", "evening"],
        "Wednesday": ["afternoon", "morning", "evening"],
        "Thursday": ["afternoon", "morning", "evening"],
        "Friday": ["afternoon", "morning", "evening"],
        "Saturday": ["afternoon", "morning", "evening"],
        "Sunday": ["afternoon", "morning", "evening"]
    },
    "Employee H": {
        "Monday": ["evening", "morning", "afternoon"],
        "Tuesday": ["evening", "morning", "afternoon"],
        "Wednesday": ["evening", "morning", "afternoon"],
        "Thursday": ["evening", "morning", "afternoon"],
        "Friday": ["evening", "morning", "afternoon"],
        "Saturday": ["evening", "morning", "afternoon"],
        "Sunday": ["evening", "morning", "afternoon"]
    },
    "Employee I": {
        "Monday": ["morning", "evening", "afternoon"],
        "Tuesday": ["morning", "evening", "afternoon"],
        "Wednesday": ["morning", "evening", "afternoon"],
        "Thursday": ["morning", "evening", "afternoon"],
        "Friday": ["morning", "evening", "afternoon"],
        "Saturday": ["morning", "evening", "afternoon"],
        "Sunday": ["morning", "evening", "afternoon"]
    },
    "Employee J": {
        "Monday": ["afternoon", "evening", "morning"],
        "Tuesday": ["afternoon", "evening", "morning"],
        "Wednesday": ["afternoon", "evening", "morning"],
        "Thursday": ["afternoon", "evening", "morning"],
        "Friday": ["afternoon", "evening", "morning"],
        "Saturday": ["afternoon", "evening", "morning"],
        "Sunday": ["afternoon", "evening", "morning"]
    }
}

for name, prefs in employee_data.items():
    scheduler.add_employee_preferences(name, prefs)

scheduler.assign_shifts()
scheduler.resolve_conflicts()
scheduler.display_schedule()
